src/server/list_family.cc

```diff
@@ -246,40 +246,38 @@ OpResult<string> OpMoveSingleShard(const OpArgs& op_args, string_view src, strin
   }
 
   quicklist* dest_ql = nullptr;
-  PrimeIterator dest_it;
-  bool new_key = false;
+  src_res->post_updater.Run();
+  DbSlice::AddOrFindResult dest_res;
   try {
-    src_res->post_updater.Run();
-    tie(dest_it, new_key) = db_slice.AddOrFind(op_args.db_cntx, dest);
+    dest_res = db_slice.AddOrFind(op_args.db_cntx, dest);
   } catch (bad_alloc&) {
     return OpStatus::OUT_OF_MEMORY;
   }
 
-  if (new_key) {
+  if (dest_res.is_new) {
     robj* obj = createQuicklistObject();
     dest_ql = (quicklist*)obj->ptr;
     quicklistSetOptions(dest_ql, GetFlag(FLAGS_list_max_listpack_size),
                         GetFlag(FLAGS_list_compress_depth));
-    dest_it->second.ImportRObj(obj);
+    dest_res.it->second.ImportRObj(obj);
 
     // Insertion of dest could invalidate src_it. Find it again.
     src_res = db_slice.FindMutable(op_args.db_cntx, src, OBJ_LIST);
     src_it = src_res->it;
     DCHECK(IsValid(src_it));
   } else {
-    if (dest_it->second.ObjType() != OBJ_LIST)
+    if (dest_res.it->second.ObjType() != OBJ_LIST)
       return OpStatus::WRONG_TYPE;
 
-    dest_ql = GetQL(dest_it->second);
-    db_slice.PreUpdate(op_args.db_cntx.db_index, dest_it);
+    dest_ql = GetQL(dest_res.it->second);
```

Comment:
I think that in this flow you run one pre update and twice post update for the source key

Reply:
The second `Run()` will do nothing, as it's "unarmed"

Reply:
(if you mean that I run it in line 249, but then I do _not_ search for it in line 265 because dest exists, and then I re-invoke `Run()` in line 279)

Reply:
yes this is what I mean. The problem is that you will run the post updater before the entry changed in line 275, so the memory accounting will be wrong.

Reply:
nice, another good catch. I think that the easiest solution would be to `FindMutable()` src again, even if no insertion took place. (which is what I did)